src/useFormStates.ts

```diff
@@ -107,6 +119,8 @@ export function useFormStates<T, I = T>(opts: UseFormStatesOpts<T, I>): { getFor
         existing[1] = input;
       }
 
+      form.readOnly = readOnlyRef.current || !!opts.readOnly;
```

Comment:
Each row is treated read-only based on an `||` of "page-level readonly" or "row-level readonly".